lib/rpc/src/tcp_client.rs

```diff
@@ -0,0 +1,463 @@
+    sockets: SocketSet<'a>,
+    server_handle: Option<SocketHandle>,
+    server_ip: IpAddress,
+    server_port: u16,
+    client_port: u16,
+    client_id: NodeId,
+    req_id: u64,
+}
+
+impl TCPClient<'_> {
+    pub fn new<'a>(
+        server_ip: IpAddress,
+        server_port: u16,
+        iface: EthernetInterface<'a, DevQueuePhy>,
+    ) -> TCPClient<'a> {
+        TCPClient {
+            iface: iface,
+            sockets: SocketSet::new(vec![]),
+            server_handle: None,
+            server_ip: server_ip,
+            server_port: server_port,
+            client_port: 10110,
+            client_id: 0,
+            req_id: 0,
+        }
+    }
+}
+
+impl ClusterClientAPI for TCPClient<'_> {
+    /// Register with controller, analogous to LITE join_cluster()
+    /// TODO: add timeout?? with error returned if timeout occurs?
+    fn join_cluster(&mut self) -> Result<NodeId, ClusterError> {
+        // create client socket
+        let tcp_rx_buffer = TcpSocketBuffer::new(vec![0; RX_BUF_LEN]);
+        let tcp_tx_buffer = TcpSocketBuffer::new(vec![0; TX_BUF_LEN]);
+        let tcp_socket = TcpSocket::new(tcp_rx_buffer, tcp_tx_buffer);
+        self.server_handle = Some(self.sockets.add(tcp_socket));
+
+        {
+            let mut socket = self.sockets.get::<TcpSocket>(self.server_handle.unwrap());
+            socket
+                .connect((self.server_ip, self.server_port), self.client_port)
+                .unwrap();
+            debug!(
+                "Attempting to connect to server {}:{}",
+                self.server_ip, self.server_port
+            );
+        }
+
+        // Connect to server
+        loop {
+            match self.iface.poll(&mut self.sockets, Instant::from_millis(0)) {
+                Ok(_) => {}
+                Err(e) => {
+                    warn!("poll error: {}", e);
+                }
+            }
+            let socket = self.sockets.get::<TcpSocket>(self.server_handle.unwrap());
+            // Waiting for send/recv forces the TCP handshake to fully complete
+            if socket.is_active() && (socket.may_send() || socket.may_recv()) {
+                debug!("Connected to server, ready to send/recv data");
+                break;
+            }
+        }
+
+        self.rpc_call(0, RPCType::Registration, Vec::new()).unwrap();
+        Ok(self.client_id)
+    }
+}
+
+/// RPC client operations
+impl RPCClientAPI for TCPClient<'_> {
+    /// calls a remote RPC function with ID
+    fn rpc_call(
+        &mut self,
+        pid: usize,
+        rpc_id: RPCType,
+        data: Vec<u8>,
+    ) -> Result<Vec<u8>, RPCError> {
+        // Create request header
+        let req_hdr = RPCHeader {
+            client_id: self.client_id,
+            pid: pid,
+            req_id: self.req_id,
+            msg_type: rpc_id,
+            msg_len: data.len() as u64,
+        };
+
+        // Serialize request header then request body
+        let mut req_data = Vec::new();
```

Comment:
We should think about a way to eliminate this allocation (ideally we should aim to have one allocation per msg sent/received)